File: picasso-transformations/src/main/java/io/github/hendraanggrian/picassotransformations/crop/CropCircleTransformation.java
package io.github.hendraanggrian.picassotransformations.crop;

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.NonNull;

import io.github.hendraanggrian.picassotransformations.Key;
import io.github.hendraanggrian.picassotransformations.internal.PaintBuilder;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class CropCircleTransformation extends CropSquareTransformation {

    public static final String TAG = "circle";

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, int height, int width, @NonNull Bitmap target) {
        final int size = Math.min(width, height);
        final Bitmap squared = super.transform(source, height, width, target);
        final Bitmap circle = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        float r = size / 2f;
        new Canvas(circle).drawCircle(r, r, r, new PaintBuilder(Paint.ANTI_ALIAS_FLAG)
                .shader(new BitmapShader(squared, BitmapShader.TileMode.CLAMP, BitmapShader.TileMode.CLAMP))
                .build());
        squared.recycle();
        return circle;
    }

    @NonNull
    @Override
    public Key getKey() {
        return Key.fromTag(TAG);
    }
}

File: pikasso/src/com/hendraanggrian/pikasso/transformation/CropCircleTransformation.kt
package com.hendraanggrian.pikasso.transformation

import android.graphics.Bitmap
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Bitmap.createBitmap
import android.graphics.BitmapShader
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Shader.TileMode.CLAMP
import com.squareup.picasso.Transformation
import kotlin.math.min

class CropCircleTransformation : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val size = min(source.width, source.height)
        val width = (source.width - size) / 2
        val height = (source.height - size) / 2
        val target = createBitmap(size, size, ARGB_8888)
        val r = size / 2f
        Canvas(target).drawCircle(r, r, r, Paint().apply {
            isAntiAlias = true
            shader = BitmapShader(source, CLAMP, CLAMP)
            // source isn't square, move viewport to center
            if (width != 0 || height != 0) shader.setLocalMatrix(Matrix().apply {
                setTranslate(-width.toFloat(), -height.toFloat())
            })
        })
        source.recycle()
        return target
    }

    override fun key() = "CropCircleTransformation"
}